File: app/src/main/java/sugtao4423/lod/MainActivity.java
package sugtao4423.lod;

import android.annotation.SuppressLint;
import android.app.AlertDialog.Builder;
import android.content.Intent;
import android.content.IntentFilter;
import android.graphics.Color;
import android.os.Build;
import android.os.Bundle;
import android.support.v4.view.PagerTabStrip;
import android.support.v4.view.ViewPager;
import android.view.View;

import java.io.File;

import sugtao4423.icondialog.IconDialog;
import sugtao4423.icondialog.IconItem;
import sugtao4423.lod.AutoLoadTLService.AutoLoadTLListener;
import sugtao4423.lod.main_fragment.Fragment_home;
import sugtao4423.lod.main_fragment.Fragment_mention;
import sugtao4423.lod.main_fragment.MainFragmentPagerAdapter;
import twitter4j.ResponseList;
import twitter4j.Status;

public class MainActivity extends LoDBaseActivity{

    private boolean resetFlag;

    private Fragment_mention fragmentMention;
    private Fragment_home fragmentHome;

    private Builder iconDialog;

    private MusicReceiver musicReceiver;

    @SuppressLint("ResourceType")
    @Override
    protected void onCreate(Bundle savedInstanceState){
        super.onCreate(savedInstanceState);
        getSupportActionBar().hide();
        setContentView(R.layout.activity_main);

        MainFragmentPagerAdapter pagerAdapter = new MainFragmentPagerAdapter(getSupportFragmentManager(), this);

        ViewPager viewPager = (ViewPager)findViewById(R.id.pager);
        viewPager.setAdapter(pagerAdapter);
        viewPager.setCurrentItem(1);
        viewPager.setOffscreenPageLimit(app.getCurrentAccount().getSelectListIds().length + 1);

        PagerTabStrip strip = (PagerTabStrip)findViewById(R.id.mainPagerTabStrip);
        strip.setTabIndicatorColor(Color.parseColor(getString(R.color.pagerTabText)));
        strip.setDrawFullUnderline(true);
        getSupportActionBar().setDisplayShowHomeEnabled(false);

        fragmentHome = pagerAdapter.getFragmentHome();
        fragmentMention = pagerAdapter.getFragmentMention();
        logIn();
        setMusicReceiver();
    }

    public void logIn(){
        if(!app.haveAccount()){
            startActivity(new Intent(this, StartOAuth.class));
            finish();
        }else{
            autoLoadTL();
        }
    }

    public void autoLoadTL(){
        if(app.getCurrentAccount().getAutoLoadTLInterval() == 0){
            return;
        }
        AutoLoadTLListener listener = new AutoLoadTLListener(){

            @Override
            public void onStatus(ResponseList<Status> statuses){
                for(Status s : statuses){
                    fragmentHome.insert(s);
                    if(app.getMentionPattern().matcher(s.getText()).find() && !s.isRetweet()){
                        fragmentMention.insert(s);
                    }
                }
            }
        };
        app.setAutoLoadTLListener(listener);
        Intent intent = new Intent(this, AutoLoadTLService.class);
        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.O){
            startForegroundService(intent);
        }else{
            startService(intent);
        }
    }

    public void setMusicReceiver(){
        musicReceiver = new MusicReceiver();
        IntentFilter filter = new IntentFilter();
        for(String s : MusicReceiver.ACTIONS_GOOGLEPLAY){
            filter.addAction(s);
        }
        for(String s : MusicReceiver.ACTIONS_SONYMUSIC){
            filter.addAction(s);
        }
        registerReceiver(musicReceiver, filter);
    }

    public void newTweet(View v){
        Intent intent = new Intent(MainActivity.this, TweetActivity.class);
        startActivity(intent);
    }

    public void option(View v){
        if(iconDialog == null){
            int black = Color.parseColor(getString(R.color.icon));
            IconItem[] items = new IconItem[6];
            items[0] = new IconItem(getString(R.string.icon_bomb).charAt(0), black, getString(R.string.tweet_bomb));
            items[1] = new IconItem(getString(R.string.icon_search).charAt(0), black, getString(R.string.search_user));
            items[2] = new IconItem(getString(R.string.icon_user).charAt(0), black, getString(R.string.account));
            items[3] = new IconItem(getString(R.string.icon_experience).charAt(0), black, getString(R.string.level_info));
            items[4] = new IconItem(getString(R.string.icon_clock).charAt(0), black, getString(R.string.use_info));
            items[5] = new IconItem(getString(R.string.icon_cog).charAt(0), black, getString(R.string.settings));
            iconDialog = new IconDialog(this).setItems(items, new OptionClickListener(this));
        }
        iconDialog.show();
    }

    public void restart(){
        resetFlag = true;
        finish();
    }

    @Override
    public void onDestroy(){
        super.onDestroy();
        unregisterReceiver(musicReceiver);
        stopService(new Intent(this, AutoLoadTLService.class));
        app.resetAccount();
        app.closeAccountDB();
        app.closeUseTimeDB();
        if(resetFlag){
            resetFlag = false;
            startActivity(new Intent(this, MainActivity.class));
        }
        clearThumbnailCache();
    }

    public void clearThumbnailCache(){
        File cache = new File(getCacheDir().getAbsolutePath() + "/web_image_cache/");
        if(!cache.exists()){
            return;
        }
        for(File f : cache.listFiles()){
            if(f.getName().startsWith("https+pbs+twimg+com+media+")){
                f.delete();
            }
        }
    }

}

File: app/src/main/java/sugtao4423/lod/MainActivity.kt
package sugtao4423.lod

import android.content.Intent
import android.os.Build
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import androidx.core.content.ContextCompat
import com.bumptech.glide.Glide
import kotlinx.coroutines.runBlocking
import sugtao4423.icondialog.IconDialog
import sugtao4423.icondialog.IconItem
import sugtao4423.lod.databinding.ActivityMainBinding
import sugtao4423.lod.main_fragment.Fragment_home
import sugtao4423.lod.main_fragment.Fragment_mention
import sugtao4423.lod.main_fragment.MainFragmentPagerAdapter
import sugtao4423.lod.ui.addaccount.AddAccountActivity
import twitter4j.ResponseList
import twitter4j.Status

class MainActivity : LoDBaseActivity() {

    private var resetFlag = false

    private lateinit var fragmentMention: Fragment_mention
    private lateinit var fragmentHome: Fragment_home

    private var iconDialog: AlertDialog.Builder? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        supportActionBar?.hide()
        val binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        if (!app.hasAccount) {
            startActivity(Intent(this, AddAccountActivity::class.java))
            finish()
            return
        }

        binding.tweetBtn.setOnClickListener { clickNewTweet() }
        binding.optionBtn.setOnClickListener { clickOption() }

        val pagerAdapter = MainFragmentPagerAdapter(supportFragmentManager, this)

        binding.pager.apply {
            adapter = pagerAdapter
            currentItem = 1
            offscreenPageLimit = app.account.selectListIds.size + 1
        }

        binding.mainPagerTabStrip.apply {
            tabIndicatorColor = ContextCompat.getColor(applicationContext, R.color.pagerTabText)
            drawFullUnderline = true
        }
        supportActionBar?.setDisplayShowHomeEnabled(false)

        fragmentHome = pagerAdapter.fragmentHome
        fragmentMention = pagerAdapter.fragmentMention
        autoLoadTL()
    }

    private fun autoLoadTL() {
        if (app.account.autoLoadTLInterval == 0) {
            return
        }
        val listener = object : AutoLoadTLService.AutoLoadTLListener {
            override fun onStatus(statuses: ResponseList<Status>) {
                statuses.map {
                    fragmentHome.insert(it)
                    if (app.mentionPattern.matcher(it.text).find() && !it.isRetweet) {
                        fragmentMention.insert(it)
                    }
                }
            }
        }
        app.autoLoadTLListener = listener
        val intent = Intent(this, AutoLoadTLService::class.java)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            startForegroundService(intent)
        } else {
            startService(intent)
        }
    }

    private fun clickNewTweet() {
        startActivity(Intent(this, TweetActivity::class.java))
    }

    private fun clickOption() {
        if (iconDialog == null) {
            val black = ContextCompat.getColor(applicationContext, R.color.icon)
            val items = arrayOf(
                    IconItem(getString(R.string.icon_bomb)[0], black, getString(R.string.tweet_bomb)),
                    IconItem(getString(R.string.icon_search)[0], black, getString(R.string.search_user)),
                    IconItem(getString(R.string.icon_user)[0], black, getString(R.string.account)),
                    IconItem(getString(R.string.icon_experience)[0], black, getString(R.string.level_info)),
                    IconItem(getString(R.string.icon_clock)[0], black, getString(R.string.use_info)),
                    IconItem(getString(R.string.icon_cog)[0], black, getString(R.string.settings))
            )
            iconDialog = IconDialog(this).setItems(items, OptionClickListener(this))
        }
        iconDialog!!.show()
    }

    fun restart() {
        resetFlag = true
        finish()
    }

    override fun onBackPressed() {
        finish()
    }

    override fun onDestroy() {
        super.onDestroy()
        stopService(Intent(this, AutoLoadTLService::class.java))
        runBlocking { app.reloadAccount() }
        if (resetFlag) {
            resetFlag = false
            startActivity(Intent(this, MainActivity::class.java))
        }
        clearImageMemoryCache()
    }

    private fun clearImageMemoryCache() {
        Glide.get(this).clearMemory()
    }

}
